### PR TITLE
add code to prevent segfault during Eval of thunk Expr

### LINKED ISSUE #32315: Evaluation of "Thunk"-Expression results in EXCEPTION_ACCESS_VIOLATION
It is probably a bit silly what I am doing. (I tried to somehow introduce a new scope block using AST and saw that an empty `let`-Block with no variables is translated to a `:thunk`-Expression, which I tried.)

    Expr(:thunk, Meta.parse("x=17"))
    Main.eval(Expr(:thunk, Meta.parse("x=17")))

This results in a crash of Julia the following error (with a fresh Julia session with version 1.1.1 on Windows 10):

```
Exception: EXCEPTION_ACCESS_VIOLATION at 0x6b5ccba9 -- jl_array_ptr_ref at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\julia.h:807 [inlined]
body_attributes at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\toplevel.c:364 [inlined]
jl_toplevel_eval_flex at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\toplevel.c:739
in expression starting at no file:0
jl_array_ptr_ref at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\julia.h:807 [inlined]
body_attributes at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\toplevel.c:364 [inlined]
jl_toplevel_eval_flex at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\toplevel.c:739
jl_toplevel_eval at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\toplevel.c:773 [inlined]
jl_toplevel_eval_in at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\toplevel.c:793
eval at .\boot.jl:328 [inlined]
eval at .\client.jl:404
jl_fptr_trampoline at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\gf.c:1842
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2197
do_call at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\interpreter.c:323
eval_value at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\interpreter.c:411
eval_stmt_value at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\interpreter.c:362 [inlined]
eval_body at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\interpreter.c:759
jl_interpret_toplevel_thunk_callback at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\interpreter.c:885
unknown function (ip: FFFFFFFFFFFFFFFE)
unknown function (ip: 00000000105E168F)
unknown function (ip: 0000000000000003)
jl_toplevel_eval_flex at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\toplevel.c:764
jl_toplevel_eval at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\toplevel.c:773 [inlined]
jl_toplevel_eval_in at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\toplevel.c:793
eval at .\boot.jl:328
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2197
eval_user_input at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.1\REPL\src\REPL.jl:85
macro expansion at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.1\REPL\src\REPL.jl:117 [inlined]
#26 at .\task.jl:259
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2197
jl_apply at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\julia.h:1571 [inlined]
start_task at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\task.c:572
Allocations: 944350 (Pool: 943830; Big: 520); GC: 1
```

Versioninfo:

```
julia> versioninfo()
Julia Version 1.1.1
Commit 55e36cc308 (2019-05-16 04:10 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i5-4570 CPU @ 3.20GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, haswell)
Environment:
  JULIA_PKGDIR = C:\.julia
```

### PR DESCRIPTION
**Pull request by ssikdar1**
# Description
Trying to address issue #32315.

I have a feeling this isn't correct, but trying to code against the situation where a `:thunk` `Expr` is passed to `Eval` with incorrect values.

For example the following now would have an error returned where they previously segfaulted.

```
julia> Main.eval(Expr(:thunk, x->x+9))
ERROR: syntax: malformed "thunk" statement
Stacktrace:
 [1] top-level scope at none:1
 [2] eval at ./boot.jl:331 [inlined]
 [3] eval(::Expr) at ./client.jl:452
 [4] top-level scope at REPL[1]:1

julia> Main.eval(Expr(:thunk, Meta.parse("x=17")))
ERROR: null stmt in body_attributes
Stacktrace:
 [1] eval at ./boot.jl:331 [inlined]
 [2] eval(::Expr) at ./client.jl:452
 [3] top-level scope at REPL[2]:1
```

Although the following does still segfault:

```
julia> Main.eval(Expr(:thunk, Meta.parse("17")))

signal (11): Segmentation fault: 11
in expression starting at none:1
body_attributes at /Users/ssikdar1/julia/src/toplevel.c:369 [inlined]
jl_toplevel_eval_flex at /Users/ssikdar1/julia/src/toplevel.c:801
jl_toplevel_eval at /Users/ssikdar1/julia/src/toplevel.c:834 [inlined]
jl_toplevel_eval_in at /Users/ssikdar1/julia/src/toplevel.c:857
eval at ./boot.jl:331 [inlined]
eval at ./client.jl:452
jl_apply at /Users/ssikdar1/julia/src/./julia.h:1691 [inlined]
do_call at /Users/ssikdar1/julia/src/interpreter.c:369
eval_stmt_value at /Users/ssikdar1/julia/src/interpreter.c:409 [inlined]
```

Closes #32315

### CODE REVIEW

**Comment by JeffBezanson** [comment]
Yes, I don't think this is correct because it doesn't actually check the structure. It checks much later, when having the wrong structure might have resulted in some null pointers, but only if you're lucky.

File: src/toplevel.c [hunk 1]
```diff
@@ -791,7 +794,11 @@ jl_value_t *jl_toplevel_eval_flex(jl_module_t *JL_NONNULL m, jl_value_t *e, int
     thk = (jl_code_info_t*)jl_exprarg(ex, 0);
     assert(jl_is_code_info(thk));
```
Comment (JeffBezanson) [inline src/toplevel.c:795]: I think it would be enough to turn this assert into an error.

Comment (ssikdar1) [reply src/toplevel.c]: Changed the assert to error in a7c56d7

**Comment by ssikdar1** [comment]
Changed the assert to error in a7c56d7.

Now all 3 of the examples no longer segfault.

```
julia> Main.eval(Expr(:thunk, Meta.parse("17")))
ERROR: malformed "thunk" statement
Stacktrace:
 [1] top-level scope at none:1
 [2] eval at ./boot.jl:331 [inlined]
 [3] eval(::Expr) at ./client.jl:452
 [4] top-level scope at REPL[1]:1

julia> Main.eval(Expr(:thunk, Meta.parse("x=17")))
ERROR: malformed "thunk" statement
Stacktrace:
 [1] top-level scope at none:1
 [2] eval at ./boot.jl:331 [inlined]
 [3] eval(::Expr) at ./client.jl:452
 [4] top-level scope at REPL[1]:1

julia> Main.eval(Expr(:thunk, x->x+9))
ERROR: malformed "thunk" statement
Stacktrace:
 [1] top-level scope at none:1
 [2] eval at ./boot.jl:331 [inlined]
 [3] eval(::Expr) at ./client.jl:452
 [4] top-level scope at REPL[2]:1
```

**Pull request merged**